content/en/blog/2023/why-histograms/index.md

```diff
@@ -0,0 +1,86 @@
+plotted for some requests over some time. From the data, you can see that 50% of
+requests are served in around 20-30ms or less, 90% of requests are served in
+under about 80ms, and 99% of requests are served in under around 90ms. You can
+very quickly see that at least 50% of your users are receiving very fast
+response times, but almost all of your users are experiencing response times
+under 90ms.
+
+![p99, p90, and p50 plotted as lines](hist-lines.png "p99, p90, and p50 plotted as a line chart with title 'response times.' Time is on the x-axis and response times in milliseconds on the y-axis. p99 response times are around 80 milliseconds. p90 response times are betweeen 60 and 80 milliseconds. p50 response times are between 20 and 30 milliseconds.")
+
+# Other metric types
+
+Another solution might be to define the SLOs you're interested in and collect
+them as non-histogram metrics in advance as gauges or counters. This approach
+works, but requires defining your SLOs before you have an understanding of your
+data distribution and requires non-trivial implementation at collection time. It
+is also inflexible; if you decide to change your SLO from 90% of requests to 99%
+of requests, you have to make and release code changes, then wait for the old
+data to age out and the new metric to collect enough data to make useful
+queries. Because histograms model data as a distribution from start to finish,
+they enable you to simply change your queries and get answers on the data you've
+already collected. Particularly with exponential histograms, arbitrary
+distribution queries can be made with very low relative error rates and minimal
+resource consumption on both the client and the analysis backend.
+
+The inflexibility of this approach also impacts your ability to gauge impact
+when your SLO is violated. For example, imagine you are collecting a gauge which
+calculates the `p99` of some metric and you define an SLO based on it. When your
+SLO is violated and an alert is triggered, how do you know it is really only
+affecting 1% of queries, 10%, or 50%? A histogram would allow you to answer that
+question by simply querying the percentiles you're interested in. You could
+collect additional gauges for each percentile, but then you've just force users
+to reimplement histograms on their own. Probably poorly.
+
+# Other data sources and metric types
+
+You may ask why you would report a separate metric rather than calculating these
+metrics from your existing log and trace data? While it is true that for _some_
+use cases, like response times, this may be possible, it is not possible for
+_all_ use cases. Even when percentiles can be calculated from existing data,
+this typically requires querying a massive amount of data to answer a very
+simple question. Further, if you are sampling your tracing and logs data, you
+may be missing crucial information required to asses whether or not you are
+meeting your SLOs.
```

Comment:
```suggestion
You may ask, "why would you report a separate metric rather than calculating these
metrics from your existing log and trace data? While it is true that for _some_
use cases, like response times, this may be possible, it is not necessarily possible for
_all_ use cases. Even when percentiles can be calculated from existing data,
this typically requires querying a massive amount of data to answer a very
simple question. Some observability backends can support this, but others cannot.
Further, if you are sampling your tracing and logs data, you miss crucial information
required to asses whether or not you are meeting your SLOs. Again, this concern is
also backend- and sampling-specific. Depending on your sampling strategy, it may
not be possible to have representative data to accurately calculate an SLO. Further,
if your backend doesn't support re-weighting counts on sampled data, it can
similarly lead to inaccurate SLOs. Using histograms is a way to avoid these subtle
problems if they apply to you.
```

Reply:
What about this:

>You may ask, "why would I report a separate metric rather than calculating it
from my existing log and trace data?" While it is true that for _some_ use
cases, like response times, this may be possible, it is not necessarily possible
for _all_ use cases. Even when quantiles can be calculated from existing data,
you may run into other problems. You need to be sure your observability backend
is able to query and analyze a large amount of existing data on-line or index
and analyze it at ingestion time. If you are sampling your logs and traces or
employing a data retention policy that ages data out, you need to be sure those
things are not affecting derived metrics, or that they are properly re-weighted,
or you risk not being able to accurately asses your SLOs. Depending on your
sampling strategy, it may not even be possible. Using histograms is a way to
avoid these subtle problems if they apply to you.

Reply:
I went out of my way when writing the whole series to try not to mention anything backend specific so I'd like to focus the wording on the underlying principles instead.

Reply:
I love that @dyladan, thanks! Let's go with that and then the post LGTM.